Работа с балансами

Kot-Mars opened the pull request:
Jira - https://faang-school.atlassian.net/browse/BJS2-66557

Commit:
chore: add migration to create balance table
Created migration for the `balance` table with fields for authorization and factual balances. Resolves: BJS2-76827

Commit:
feat: added OneToOne relationship between Account and Balance entities
-Create Entity Balance
- Added a OneToOne relationship in the Account entity via the balance field.
- The relationship is set using @OneToOne(mappedBy = "account") in the Account entity. Addresses: BJS2-66557

PavelEzhkov committed:
Merge pull request #454 from CorporationX/BJS2-76927

Cоздания таблицы balance

Commit:
Merge remote-tracking branch 'origin/BJS2-66557' into BJS2-66557

Commit:
BJS2-66557: Создание всей логики
Добавлен Entity balance, добавление в миграцию таблиц генерация айди таблицы, создание сервис логики(создание баланса, обновление баланса и найти баланс по номеру аккаунта), сделан контроллер и добавлен обработчик ошибок handle
(GlobalExceptionHandler). Добавлен метод в аккаунт репозиторий

Kot-Mars committed:
BJS2-66557: Create Test
Create positive and negative test BalanceServiceImplTest

antonykapitukov commented on src/main/java/faang/school/accountservice/controller/BalanceController.java:
```diff
@@ -0,0 +1,46 @@
+@RestController
+@RequestMapping("/api/v1/balance")
+public class BalanceController {
+    private final BalanceServiceImpl balanceService;
```
рекомендую инжектировать не конкретную реализацию, а интерфейс, чтобы:
 • соблюдать принцип инверсии зависимостей (D – Dependency Inversion Principle) из SOLID,
 • обеспечить возможность подмены реализации (например, мок-объекты при тестировании),
 • повысить гибкость и масштабируемость приложения в будущем.

antonykapitukov commented on src/main/java/faang/school/accountservice/service/balance/BalanceServiceImpl.java:
```diff
@@ -0,0 +1,89 @@
+
+    @Override
+    @Transactional
+    public void updateBalance(BalanceRequestDto request) {
```
рекомендую чтобы метод возвращал обновленную дто, чтобы отследить корректность выполнения метода и передать ее в контроллер

antonykapitukov commented on src/main/java/faang/school/accountservice/service/balance/BalanceServiceImpl.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    @Override
+    public void createBalance(BalanceRequestDto request) {
```
рекомендую чтобы метод возвращал созданную дто, чтобы отследить корректность выполнения метода и передать ее в контроллер

antonykapitukov commented on src/main/java/faang/school/accountservice/controller/BalanceController.java:
```diff
@@ -0,0 +1,46 @@
+    public BalanceRequestDto createBalance(@RequestBody BalanceRequestDto request) {
+        balanceService.createBalance(request);
+        log.info("Created new balance: {}", request);
+        return request;
```
не очень логично что эндпоинт контроллера возвращает request (запрос) а не response (ответ)

antonykapitukov replied:
предлагаю создать отдельную дтошку для ответа например BalanceResponseDto

antonykapitukov commented on src/main/java/faang/school/accountservice/controller/BalanceController.java:
```diff
@@ -0,0 +1,46 @@
+    public BalanceRequestDto updateBalance(@RequestBody BalanceRequestDto request) {
+        balanceService.updateBalance(request);
+        log.info("Updated balance: {}", request);
+        return request;
```
не очень логично что эндпоинт контроллера возвращает request (запрос) а не response (ответ)

antonykapitukov commented on src/main/java/faang/school/accountservice/controller/BalanceController.java:
```diff
@@ -0,0 +1,46 @@
+    public BalanceResponseDto getBalance(@PathVariable String accountNumber) {
+       BalanceResponseDto balanceResponseDto = balanceService.getBalanceByAccountNumber(accountNumber);
+       log.info("Get balance for account number: {}", accountNumber);
+       return balanceResponseDto;
```
предлагаю вот так же сделать и в предыдущих методах

antonykapitukov commented on src/main/java/faang/school/accountservice/dto/BalanceRequestDto.java:
```diff
@@ -0,0 +1,16 @@
+@Data
+public class BalanceRequestDto {
+    @NotNull
+    private String accountNumber;
```
можно добавить еще аннотации валидации например NotEmpty

antonykapitukov commented on src/main/java/faang/school/accountservice/dto/BalanceRequestDto.java:
```diff
@@ -0,0 +1,16 @@
+    @NotNull
+    private String accountNumber;
+    @NotNull
+    private BigDecimal authorizationBalance;
```
можно добавить еще аннотации валидации например чтобы не было отрицательное значение

antonykapitukov commented on src/main/java/faang/school/accountservice/dto/BalanceRequestDto.java:
```diff
@@ -0,0 +1,16 @@
+
+@Data
+public class BalanceRequestDto {
+    @NotNull
```
можно добавить сообщение которые будет выбрасываться если поле будет null, например message = "Account Number can't be null"

antonykapitukov commented on src/main/java/faang/school/accountservice/dto/BalanceRequestDto.java:
```diff
@@ -0,0 +1,16 @@
+public class BalanceRequestDto {
+    @NotNull
+    private String accountNumber;
+    @NotNull
```
тут тоже можно добавить сообщение

antonykapitukov commented on src/main/java/faang/school/accountservice/dto/BalanceRequestDto.java:
```diff
@@ -0,0 +1,16 @@
+    @NotNull
+    private BigDecimal authorizationBalance;
+    @NotNull
+    private BigDecimal factualBalance;
```
тут все аналогично

antonykapitukov commented on src/main/java/faang/school/accountservice/dto/BalanceResponseDto.java:
```diff
@@ -0,0 +1,23 @@
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class BalanceResponseDto {
```
зачем дтошке аннотации AllArgsConstructor NoArgsConstructor Builder?

SergeyErmoshenko replied:
В тестах удобнее, чтобы не через конструктор

antonykapitukov commented on src/main/java/faang/school/accountservice/entity/Balance.java:
```diff
@@ -0,0 +1,62 @@
+import org.hibernate.annotations.CreationTimestamp;
+import org.hibernate.annotations.UpdateTimestamp;
+
+@Data
```
Лучше использовать более частичные аннотации, такие как @Getter и @Setter, а equals и hashCode реализовать вручную с учётом особенностей JPA. Это позволит более точно описать логику равенства и избежит потенциальных проблем при работе с ORM. Неожиданные реализации toString Метод toString, создаваемый с помощью @Data, может подтягивать лениво загружаемые свойства, что приводит к дополнительным запросам в базу и ухудшению производительности.

SergeyErmoshenko replied:
В консоль ты ентити выводить не будешь, а вот хэш код и иквалс для них нужно самим переопределить

antonykapitukov commented on src/main/java/faang/school/accountservice/entity/Balance.java:
```diff
@@ -0,0 +1,62 @@
+    @Column(name = "balance_id")
+    private Long balanceId;
+
+    @NotNull
```
Аннотации валидации (@NotNull и т.п.) лучше применять в DTO, отвечающих за прием входящих данных, а не непосредственно в сущностях. Это помогает отделить слой валидации запросов от бизнес-логики и моделей данных, упрощая поддержку и тестирование. В сущностях же следует держать только маппинг и ограничения, необходимые для работы с БД.

antonykapitukov replied:
в остальных полях тоже

antonykapitukov commented on src/main/java/faang/school/accountservice/handler/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,64 @@
+                .property("service", "account")
+                .build();
+        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(error);
+    }
```
красиво выглядит

antonykapitukov commented on src/main/java/faang/school/accountservice/handler/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,64 @@
+                .detail("Please verify that the account number is correct and try again.")
+                .property("service", "account")
+                .build();
+        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(error);
```
статус я бы вынес в аннотацию

antonykapitukov commented on src/main/java/faang/school/accountservice/mapper/BalanceMapper.java:
```diff
@@ -0,0 +1,23 @@
+    Balance toBalance(BalanceResponseDto balanceResponseDto);
+
+    @Mapping(target = "account", ignore = true)
+    Balance toBalance(BalanceRequestDto balanceRequestDto);
```
я бы на каждую операцию (обновление, создание и тд) сделал бы свою дтошку

Kot-Mars replied:
Не вижу смысла создавать dto для обновления и создания т.к. они будут одинаковые и это будет уже нарушение DRY (Don't Repeat Yourself)

SergeyErmoshenko replied:
Может он имел ввиду отдельную дто на обновление, я бы добавил ее

antonykapitukov commented on src/main/java/faang/school/accountservice/service/balance/BalanceServiceImpl.java:
```diff
@@ -0,0 +1,89 @@
+    public void createBalance(BalanceRequestDto request) {
+        Account account = getAccount(request.getAccountNumber());
+
+        if (account.getBalance() != null) {
```
логику проверки я бы вынес в отдельный метод

antonykapitukov commented on src/main/java/faang/school/accountservice/service/balance/BalanceServiceImpl.java:
```diff
@@ -0,0 +1,89 @@
+        balance.setAccount(account);
+        balance.setAuthorizationBalance(request.getAuthorizationBalance());
+        balance.setFactualBalance(request.getFactualBalance());
+        balance.setCreatedAt(LocalDateTime.now());
```
Balance balance = Balance.builder()
         .account(account)
         .authorizationBalance(request.getAuthorizationBalance())
         .factualBalance(request.getFactualBalance())
         .createdAt(LocalDateTime.now())
         .build();

antonykapitukov commented on src/test/java/faang/school/accountservice/service/balance/BalanceServiceImplTest.java:
```diff
@@ -0,0 +1,168 @@
+        balance.setFactualBalance(BigDecimal.valueOf(100));
+    }
+
+    //Positive
```
коммент лишний

Kot-Mars replied:
Почему лишний? Тут отделение позитивных тестов, от негативных

antonykapitukov commented on src/test/java/faang/school/accountservice/service/balance/BalanceServiceImplTest.java:
```diff
@@ -0,0 +1,168 @@
+
+        balance.setAccount(account);
+        balance.setAuthorizationBalance(BigDecimal.valueOf(90));
+        balance.setFactualBalance(BigDecimal.valueOf(100));
```
у сущности Balance есть аннотация Buider, это можно переписать с помощью билдера

SergeyErmoshenko commented on src/main/java/faang/school/accountservice/entity/Account.java:
```diff
@@ -70,4 +72,7 @@ public class Account {
     @Column(name = "version")
     private Long version;
+
+    @OneToOne(mappedBy = "account", fetch = FetchType.LAZY)
```
Давай название явно укажем

SergeyErmoshenko commented on src/main/java/faang/school/accountservice/service/balance/BalanceServiceImpl.java:
```diff
@@ -0,0 +1,89 @@
+            balance.setFactualBalance(request.getFactualBalance());
+            log.info("Updating balance for account {}", request.getAccountNumber());
+            balanceRepository.save(balance);
+        } catch (OptimisticLockException e) {
```
А ретраи будут?

SergeyErmoshenko commented on src/main/resources/db/changelog/changeset/V002__account-service_create_table_balance.sql:
```diff
@@ -0,0 +1,14 @@
+        REFERENCES payment_account(account_number)
+);
+
+CREATE INDEX IF NOT EXISTS idx_balance_account_number ON  balance(account_number);
```
Пробел лишний после ON

Commit:
BJS2-66557: Изменение ошибок
Методы BalanceService теперь возвращают BalanceResponseDto, добавление аннотаций в BalanceResponseDto, изменение контролера BalanceController

Kot-Mars committed:
BJS2-66557: Исправление ошибок

PavelEzhkov commented on src/main/java/faang/school/accountservice/controller/BalanceController.java:
```diff
@@ -0,0 +1,46 @@
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public BalanceResponseDto createBalance(@RequestBody BalanceRequestDto request) {
```
BalanceRequestDto  не валидируется `@Valid`

PavelEzhkov commented on src/main/java/faang/school/accountservice/controller/BalanceController.java:
```diff
@@ -0,0 +1,46 @@
+        return response;
+    }
+
+    @PutMapping("/update")
```
Путь избыточен (PUT уже подразумевает обновление)

PavelEzhkov commented on src/main/java/faang/school/accountservice/entity/Balance.java:
```diff
@@ -0,0 +1,55 @@
+import jakarta.persistence.SequenceGenerator;
+import jakarta.persistence.Table;
+import jakarta.persistence.Version;
+import lombok.*;
```
Разверни импорт

PavelEzhkov commented on src/main/java/faang/school/accountservice/handler/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,64 @@
+public class GlobalExceptionHandler {
+    @ExceptionHandler(AccountNotFoundException.class)
+    public ResponseEntity<ErrorResponse> handleAccountNotFondException(AccountNotFoundException e) {
+        ErrorResponse error = ErrorResponse.builder(e, HttpStatus.NOT_FOUND, e.getMessage())
```
Возврат полного сообщения исключения (e.getMessage()) может раскрыть внутреннюю логику

PavelEzhkov commented on src/main/java/faang/school/accountservice/handler/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,64 @@
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@RestControllerAdvice
+public class GlobalExceptionHandler {
```
Я бы добавил везде логирование

PavelEzhkov commented on src/main/java/faang/school/accountservice/service/balance/BalanceServiceImpl.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    @Override
+    public BalanceResponseDto createBalance(BalanceRequestDto request) {
```
@Transactional

PavelEzhkov commented on src/main/java/faang/school/accountservice/service/balance/BalanceServiceImpl.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    @Override
+    public BalanceResponseDto getBalanceByAccountNumber(String accountNumber) {
```
@Transactional

Kot-Mars committed:
BJS2-66557: Исправление ошибок
Добавление логов в GlobalExceptionHandler,
добавление @Transactional в BalanceServiceImpl
Добавление @Valid в BalanceController
